xapian-core/cluster/lcd_clusterer.cc

```diff
@@ -0,0 +1,164 @@
+    bool operator() (const pair<PSet::iterator, double> &a,
+		     const pair<PSet::iterator, double> &b) const {
+	return a.second < b.second;
+    }
+};
+
+ClusterSet
+LCDClusterer::cluster(const MSet &mset)
+{
+    LOGCALL(API, ClusterSet, "LCDClusterer::cluster", mset);
+
+    doccount size = mset.size();
+    unsigned int k_ = k;
+    if (k_ >= size)
+	k_ = size;
+
+    // Store each document and its rel score from given mset
+    set<pair<Point, double>, pcompare> points;
+
+    // Initiliase points
+    TermListGroup tlg(mset);
+    for (MSetIterator it = mset.begin(); it != mset.end(); ++it)
+	points.emplace(Point(tlg, it.get_document()), it.get_weight());
+
+    // Container for holding the clusters
+    ClusterSet cset;
+
+    // Instantiate this for computing distance between points
+    CosineDistance distance;
+
+    // First cluster center
+    PSet::iterator cluster_center = points.begin();
+
+    for (unsigned int cnum = 1; cnum <= k_; ++cnum) {
+	// Container for new cluster
+	Cluster new_cluster;
+
+	// The original algorithm accepts a parameter 'k' which is the number
+	// of documents in each cluster, which can be hard to tune and
+	// especially when using this in conjunction with the diversification
+	// module. So, for now LCD clustering accepts 'k' as the number of
+	// clusters and divides the documents equally in the first k-1 clusters
+	// and the remaining in the last cluster. This needs to be tested on a
+	// dataset to see how well this works.
+	unsigned int num_points = size / k_;
+	if (cnum == k_)
+	    num_points = size - (k_ - 1) * (size / k_);
+
+	/* Select (num_points - 1) nearest points to cluster_center from
+	*  from 'points' and form a new cluster
+	*/
+
+	// Store distances of each point from current cluster center
+	// Iterator of each point is stored for fast deletion from 'points'
+	vector<pair<PSet::iterator, double> > dist_vector;
+	for (auto it = points.begin(); it != points.end(); ++it) {
+	    if (it == cluster_center)
+		continue;
+
+	    double dist = distance.similarity(cluster_center->first, it->first);
+	    dist_vector.push_back(make_pair(it, dist));
+	}
+
+	// Sort dist_vector in ascending order of distance
+	sort(dist_vector.begin(), dist_vector.end(), dcompare());
+
+	// Add first num_points-1 to cluster
+	for (unsigned int i = 0; i < num_points - 1; ++i) {
+	    auto piterator = dist_vector[i].first;
+	    // Add to cluster
+	    new_cluster.add_point(piterator->first);
+	    // Remove from 'points'
+	    points.erase(piterator);
+	}
+
+	// Now select a new cluster center which is the point that is
+	// farthest away from the current cluster center
+	double max_dist = -1;
+	PSet::iterator new_cluster_center;
+	for (auto it = points.begin(); it != points.end(); ++it) {
+	    if (it == cluster_center)
+		continue;
+
+	    double dist = distance.similarity(cluster_center->first, it->first);
+	    if (dist > max_dist) {
+		max_dist = dist;
+		new_cluster_center = it;
+	    }
+	}
```

Comment:
Can't we just use `dist_vector.back().first` here to find the new centre?

Reply:
You're right, this step is unnecessary. I've pushed a commit with the suggested change, thanks.